lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -832,5 +889,25 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
       return new DefaultBulkScorer(new ConstantScoreScorer(this, 0f, ScoreMode.COMPLETE_NO_SCORES, disi));
     }
 
+    // Perform a cache load asynchronously
+    private void cacheAsynchronously(LeafReaderContext context, IndexReader.CacheHelper cacheHelper) {
+      /*
+       * If the current query is already being asynchronously cached,
+       * do not trigger another cache operation
+       */
+      if (inFlightAsyncLoadQueries.add(in.getQuery()) == false) {
+        return;
+      }
+
+      FutureTask<Void> task = new FutureTask<>(() -> {
+        DocIdSet localDocIdSet = cache(context);
+        putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
```

Comment:
Can you save the return value from `putIfAbsent` call and then `assert` it's `null`?  I.e. the cache should never already have this query since we checked up-front in thread tight manner?

Reply:
`putIfAbsent` does not return anything today -- changing that would be an orthogonal change IMO. Maybe as a follow up?

Reply:
Ahh OK let's do that later then.